### PR TITLE
[java] Move NPathComplexity from metrics to design

### PR DESCRIPTION
**Pull request by oowekyala**
See #709 
We still need to move Cyclomatic Complexity into design too

### CODE REVIEW

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/design/NPathComplexityRule.java [hunk 1]
```diff
@@ -4,262 +4,52 @@
...
-    }
-
-    @Override
-    public Object visit(ASTDoStatement node, Object data) {
-        // (npath of do + bool_comp of do + 1) * npath of next
-
-        int boolCompDo = sumExpressionComplexity(node.getFirstChildOfType(ASTExpression.class));
-
-        Integer nPathDo = (Integer) ((JavaNode) node.getFirstChildOfType(ASTStatement.class)).jjtAccept(this, data);
+    private static final IntegerProperty REPORT_LEVEL_DESCRIPTOR = new IntegerProperty(
```
Comment (adangel) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/design/NPathComplexityRule.java]: We have now a different property (well, the name is different). Assume, someone uses the rule in a custom rule, such as:

``` 
    <rule ref="java-codesize/NPathComplexity">
        <properties>
            <property name="minimum" value="100.0" />
        </properties>
    </rule>
```

We would now fail executing the ruleset with the error "Cannot set non-existant property 'minimum' on Rule NPathComplexity".

We could add backwards compatibility, by defining the old property additionally, e.g.:

`private static final DoubleProperty MINIMUM_DESCRIPTOR = new DoubleProperty("minimum", "Deprecated! Minimum reporting threshold", 0d, 300d, 200d,
            2.0f);`

And get the actual value by this way:

```
    private int getReportLevel() {
        double oldProp = getProperty(MINIMUM_DESCRIPTOR);
        if (oldProp != 200.0) {
            System.err.println("Warning! Rule NPathComplexity uses deprecated property 'minimum'. Future versions of PMD will remove support for this property. Please use 'reportLevel' instead!");
            return (int)oldProp;
        }

        return getProperty(REPORT_LEVEL_DESCRIPTOR);
    }
```

Is this worth it? The user might now update to 6.0.0 and it "just works" - not sure, if the warning is ever noticed... and with 7.0.0 it suddenly fails.... Maybe we just move the failure one release into the future :)

I've just tried it, it's at least possible...

In any case, I'll write a note into the release notes, that the rules has been revamped and the property is different now, before I'll merge it.

@jsotuyod Do you think, such a backwards-compatibility is necessary?

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/design/NPathComplexityRule.java [hunk 2]
```diff
@@ -4,262 +4,52 @@
...
 
-        return Integer.valueOf(boolCompDo + nPathDo + 1);
-    }
-
-    @Override
-    public Object visit(ASTForStatement node, Object data) {
-        // (npath of for + bool_comp of for + 1) * npath of next
 
-        int boolCompFor = sumExpressionComplexity(node.getFirstDescendantOfType(ASTExpression.class));
+    private static int reportLevel = 200;
```
Comment (adangel) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/design/NPathComplexityRule.java]: This doesn't need to be static.

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/design/NPathComplexityRule.java (hunk 1 above)
Comment (jsotuyod) [reply pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/design/NPathComplexityRule.java]: I love and totally support this idea!

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/design/NPathComplexityRule.java [hunk 3]
```diff
@@ -4,262 +4,52 @@
...
-
-    @Override
-    public Object visit(ASTDoStatement node, Object data) {
-        // (npath of do + bool_comp of do + 1) * npath of next
-
-        int boolCompDo = sumExpressionComplexity(node.getFirstChildOfType(ASTExpression.class));
-
-        Integer nPathDo = (Integer) ((JavaNode) node.getFirstChildOfType(ASTStatement.class)).jjtAccept(this, data);
+    private static final IntegerProperty REPORT_LEVEL_DESCRIPTOR = new IntegerProperty(
+        "reportLevel", "N-Path Complexity reporting threshold", 1, 30, 200, 1.0f);
```
Comment (jsotuyod) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/design/NPathComplexityRule.java]: If I'm reading this correctly, the property has a minimum value of 1, a maximum of 30, and a default of 200??

Something doesn't sound right about that...

Comment (oowekyala) [reply pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/design/NPathComplexityRule.java]: Hmm curiously the descriptor doesn't fail during construction... We should add a test case to the numeric descriptors and correct that

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/design/NPathComplexityRule.java (hunk 1 above)
Comment (oowekyala) [reply pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/design/NPathComplexityRule.java]: Making it compatible and logging a warning is probably the smoothest we can do so +1

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/design/NPathComplexityRule.java (hunk 3 above)
Comment (oowekyala) [reply pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/design/NPathComplexityRule.java]: I added #725 to track that

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/design/NPathComplexityRule.java [hunk 4]
```diff
@@ -20,33 +24,53 @@
...
+        reportLevel = getReportLevel();
 
         super.visit(node, data);
         return data;
     }
 
 
+    private int getReportLevel() {
+        double oldProp = getProperty(MINIMUM_DESCRIPTOR);
+        if (oldProp != 200.0) {
```
Comment (jsotuyod) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/design/NPathComplexityRule.java]: `oldProp != MINIMUM_DESCRIPTOR.defaultValue()`?

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/design/NPathComplexityRule.java [hunk 5]
```diff
@@ -20,33 +24,53 @@
  */
 public class NPathComplexityRule extends AbstractJavaMetricsRule {
 
-    private static final IntegerProperty REPORT_LEVEL_DESCRIPTOR = new IntegerProperty(
-        "reportLevel", "N-Path Complexity reporting threshold", 1, 30, 200, 1.0f);
+    private static final Logger LOG = Logger.getLogger(NPathComplexityRule.class.getName());
+
+    private static final DoubleProperty MINIMUM_DESCRIPTOR
```
Comment (jsotuyod) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/design/NPathComplexityRule.java]: mark this as `@deprecated`, so we can remember to get rid of this in 7.0.0

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/design/NPathComplexityRule.java (hunk 3 above)
Comment (adangel) [reply pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/design/NPathComplexityRule.java]: Btw: the original MINIMUM_DESCRIPTOR had also a range 0..100, and we used 200 as default :)

**Pull request merged**